### ISSUE TITLE
Dumping two processes connected by pipe fails

### ISSUE THREAD

**Issue by Cygon** [checkpoint-restore/criu#2335]
**Description**
I am trying to dump a running ffmpeg + aomenc pair, two processes that are connected via a pipe (one's `stdout` is the other's `stdin`)

In order to force a common parent for both processes, I launched them from a shell script, then ran said shell script via `setsid ./run.sh &> stdout.log &`

The result in `pstree` looks like this:
```
# pstree
init─┬─(other things)
     ├─run.sh─┬─aom-av1-lavish-───31*[{aom-av1-lavish-}]
     │        └─ffmpeg───16*[{ffmpeg}]

```

When I now try to dump the process id of `run.sh`:
```
# criu dump \
        --tree 13626 \
        --work-dir "/home/cygon/criutest/" \
        --images-dir "/home/cygon/criutest/state" \
        --skip-mnt "/mnt" \
        --file-validation 'filesize' \
        --log-file "criu-dump.log" \
        -vv
```

I get an error message (see below)

**Steps to reproduce the issue:**
1. Write a shell script that runs two processes connected via a pipe
2. Try to suspend the PID of the shell script

**Describe the results you received:**
I get the following error message in my `criu-dump.log`
```
(02.903063) Dumping pstree (pid: 13626)
(02.903066) ----------------------------------------
(02.903070) Process: 13626(13626)
(02.903117) Process: 13627(13627)
(02.903122) Process: 13628(13628)
(02.903128) ----------------------------------------
(02.903219) cg: All tasks in criu's cgroups. Nothing to dump.
(02.903223) unix: Dumping external sockets
(02.903296) Error (criu/tty.c:409): tty: Found slave peer index 0 without correspond master peer
```

**Describe the results you expected:**
Since I merely used `/bin/sh` as a wrapper to become the parent of the two piped processes, I was hoping that this tree of 3 processes could be dumped.

**CRIU logs and information:**
[criu-dump.log.gz](https://github.com/checkpoint-restore/criu/files/14006960/criu-dump.log.gz)

<details><summary>Output of `criu --version`:</summary>
<p>

```
Version: 3.18
```

</p>
</details>

<details><summary>Output of `criu check --all`:</summary>
<p>

```
arn  (criu/cr-check.c:822): Dirty tracking is OFF. Memory snapshot will not work.
Warn  (criu/cr-check.c:1161): Loginuid restore is OFF.
Error (criu/cr-check.c:1221): UFFD is not supported
Error (criu/cr-check.c:1221): UFFD is not supported
Warn  (criu/cr-check.c:1344): Nftables based locking requires libnftables and set concatenations support
Looks good but some kernel features are missing
which, depending on your process tree, may cause
dump or restore failure.
```

</p>
</details>

**Additional environment details:**
I am successfully dumping/restoring standalone `ffmpeg` processes and `aomenc` processes with no problems, some having been launched around last December. It's just that the scenario with a pipe (or perhaps `/bin/sh` as a parent process?) gives me problems.

**Comment by adrianreber** [comment]
Did you try checkpointing and restoring with `-j` (`--shell-job`)?

**Comment by adrianreber** [comment]
Or try `setsid ./run.sh &> stdout.log < /dev/null &` to disconnect FD 0 also.

**Comment by Cygon** [comment]
Disconnecting `stdin` hasn't helped.

I have tried a few other things in the meantime and I kept running into issues whenever a shell script was involved. In the end, I solved it by writing a small C program that forks two times to spawn `ffmpeg` and `SvtAv1EndApp`, piping the former's `stdout` through itself into the latter's `stdin` pipe

This neatly produces a process tree where the C program is the parent of both `ffmpeg` and `SvtAv1EndApp`:

```
init─┬─NuclexCriuEncod─┬─SvtAv1EncApp───113*[{SvtAv1EncApp}]
     │                 └─ffmpeg───176*[{ffmpeg}]
```

I run it with 

```
setsid "$executable" \
        &> stdout.log \
        &
```

I have no idea what my shell (bash 5.1.16) does differently, but apart from having to compile a small executable for each encode job, I'm mostly happy with this solution.
